File: cses/04_graph_algorithms/1695_police-chase/solution.py
import sys
from collections import deque


def main():
    data = sys.stdin.buffer.read().split()
    pos = 0
    n = int(data[pos]); pos += 1
    m = int(data[pos]); pos += 1

    # Danh sách cung cho mạng luồng.
    to = []
    cap = []
    graph = [[] for _ in range(n + 1)]
    edges = []  # các con đường vô hướng gốc (a, b)

    def add_edge(u, v, c):
        # Cạnh vô hướng: cả hai chiều đều có capacity c, mỗi chiều là reverse
        # (residual) của chiều kia. Chỉ số liên tiếp nên eid ^ 1 là cung ngược.
        graph[u].append(len(to)); to.append(v); cap.append(c)
        graph[v].append(len(to)); to.append(u); cap.append(c)

    # Đọc các con đường và dựng mạng luồng với capacity 1 cho mỗi cạnh.
    for _ in range(m):
        a = int(data[pos]); pos += 1
        b = int(data[pos]); pos += 1
        edges.append((a, b))
        add_edge(a, b, 1)

    s, t = 1, n

    level = [-1] * (n + 1)
    it = [0] * (n + 1)

    # BFS phân tầng: gán level theo khoảng cách từ s trong residual graph.
    def bfs():
        for i in range(n + 1):
            level[i] = -1
        level[s] = 0
        q = deque([s])
        while q:
            u = q.popleft()
            for eid in graph[u]:
                v = to[eid]
                if cap[eid] > 0 and level[v] < 0:
                    level[v] = level[u] + 1
                    q.append(v)
        return level[t] >= 0

    # DFS đẩy luồng theo blocking flow trên đồ thị phân tầng.
    def dfs(u, f):
        if u == t:
            return f
        gu = graph[u]
        while it[u] < len(gu):
            eid = gu[it[u]]
            v = to[eid]
            if cap[eid] > 0 and level[v] == level[u] + 1:
                d = dfs(v, f if f < cap[eid] else cap[eid])
                if d > 0:
                    # Cập nhật residual: giảm cung xuôi, tăng cung ngược.
                    cap[eid] -= d
                    cap[eid ^ 1] += d
                    return d
            it[u] += 1
        return 0

    sys.setrecursionlimit(10000)

    # Chạy Dinic: lặp BFS phân tầng rồi đẩy blocking flow tới khi hết luồng.
    INF = float('inf')
    while bfs():
        for i in range(n + 1):
            it[i] = 0
        while dfs(s, INF) > 0:
            pass

    # Tập đỉnh còn tới được từ s trong residual graph -> một phía của min cut.
    visited = [False] * (n + 1)
    visited[s] = True
    q = deque([s])
    while q:
        u = q.popleft()
        for eid in graph[u]:
            v = to[eid]
            if cap[eid] > 0 and not visited[v]:
                visited[v] = True
                q.append(v)

    # Con đường cần đóng: cạnh có đúng một đầu thuộc tập S (nằm trên lát cắt).
    cut = [(a, b) for (a, b) in edges if visited[a] != visited[b]]

    out = [str(len(cut))]
    out.extend("%d %d" % (a, b) for (a, b) in cut)
    sys.stdout.write("\n".join(out) + "\n")


if __name__ == "__main__":
    main()

File: cses/04_graph_algorithms/1695_police-chase/solution.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <queue>
#include <utility>
#include <vector>

using namespace std;

// Dinic cho đồ thị vô hướng: mỗi cạnh gốc thành hai cung ngược chiều capacity 1,
// hai cung là reverse edge của nhau (chỉ số liên tiếp nên edge ^ 1 là cung ngược).
class UndirectedDinic {
public:
    explicit UndirectedDinic(int vertexCount)
        : graph(vertexCount + 1), level(vertexCount + 1), nextEdge(vertexCount + 1) {}

    // Thêm một con đường vô hướng: hai cung a->b và b->a, mỗi cung capacity 1.
    void addEdge(int first, int second) {
        graph[first].push_back(static_cast<int>(destination.size()));
        destination.push_back(second);
        capacity.push_back(1);
        graph[second].push_back(static_cast<int>(destination.size()));
        destination.push_back(first);
        capacity.push_back(1);
    }

    // Tính max-flow từ source đến sink bằng Dinic; giá trị này bằng min cut.
    long long maximumFlow(int source, int sink) {
        long long result = 0;
        while (buildLevels(source, sink)) {
            fill(nextEdge.begin(), nextEdge.end(), 0);
            while (const long long pushed =
                       sendFlow(source, sink, numeric_limits<long long>::max() / 4)) {
                result += pushed;
            }
        }
        return result;
    }

    // BFS trên residual graph (chỉ theo cung còn cap > 0): tập S của min cut.
    vector<char> reachableFrom(int source) const {
        vector<char> visited(graph.size(), false);
        queue<int> pending;
        visited[source] = true;
        pending.push(source);
        while (!pending.empty()) {
            const int vertex = pending.front();
            pending.pop();
            for (const int edge : graph[vertex]) {
                const int next = destination[edge];
                if (capacity[edge] > 0 && !visited[next]) {
                    visited[next] = true;
                    pending.push(next);
                }
            }
        }
        return visited;
    }

private:
    vector<vector<int>> graph;
    vector<int> destination;
    vector<long long> capacity;
    vector<int> level;
    vector<size_t> nextEdge;

    // BFS phân tầng: gán level theo khoảng cách từ source trong residual graph.
    bool buildLevels(int source, int sink) {
        fill(level.begin(), level.end(), -1);
        queue<int> pending;
        level[source] = 0;
        pending.push(source);
        while (!pending.empty()) {
            const int vertex = pending.front();
            pending.pop();
            for (const int edge : graph[vertex]) {
                const int next = destination[edge];
                if (capacity[edge] > 0 && level[next] == -1) {
                    level[next] = level[vertex] + 1;
                    pending.push(next);
                }
            }
        }
        return level[sink] != -1;
    }

    // DFS đẩy luồng dọc theo blocking flow trên đồ thị phân tầng.
    long long sendFlow(int vertex, int sink, long long pushed) {
        if (vertex == sink || pushed == 0) {
            return pushed;
        }
        for (size_t& index = nextEdge[vertex]; index < graph[vertex].size(); ++index) {
            const int edge = graph[vertex][index];
            const int next = destination[edge];
            if (capacity[edge] == 0 || level[next] != level[vertex] + 1) {
                continue;
            }
            const long long sent = sendFlow(next, sink, min(pushed, capacity[edge]));
            if (sent > 0) {
                // Cập nhật residual: giảm cung xuôi, tăng cung ngược (edge ^ 1).
                capacity[edge] -= sent;
                capacity[edge ^ 1] += sent;
                return sent;
            }
        }
        return 0;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;
    UndirectedDinic network(n);
    // Đọc và lưu các con đường vô hướng, đồng thời dựng mạng luồng capacity 1.
    vector<pair<int, int>> streets(m);
    for (auto& [first, second] : streets) {
        cin >> first >> second;
        network.addEdge(first, second);
    }

    // Max-flow từ ngân hàng (1) đến bến cảng (n) = số con đường tối thiểu phải đóng.
    network.maximumFlow(1, n);
    const vector<char> reachable = network.reachableFrom(1);
    // Con đường cần đóng là cạnh có đúng một đầu thuộc S (nằm trên lát cắt).
    vector<pair<int, int>> cut;
    for (const auto& [first, second] : streets) {
        if (reachable[first] != reachable[second]) {
            cut.emplace_back(first, second);
        }
    }

    cout << cut.size() << '\n';
    for (const auto& [first, second] : cut) {
        cout << first << ' ' << second << '\n';
    }
    return 0;
}
